specs/introduction.md

```diff
@@ -13,3 +13,63 @@ To simplify we separate our proofs into two components.
 2. **EVM proof**: This checks that the correct opcode is called at the correct time. It checks the validity of these opcodes. It also confirms that for each of these opcodes the state proof performed the correct operation.
 
 Only after verifying both proofs are we confident that that Ethereum block is executed correctly.
+
+
+## Bus mapping
+
+
+Our state proof and EVM proof need to talk to eachother. We need some kind of accumulator to do this. Merkle trees where for a long time the best way of achiveing such communication. But this would add 1000’s of consraints to each opcode. instead we use a plookup key value mapping. We call this bus mapping afte the bus on a traditional computer.
+
+One of the big challenges of building a snark to verify a VM is that you need to be able to read random values / opcodes at any time. To do this we use key-value mappings. Key value mappings are basically random linear combinations of all elements
+
+
+### Plookup key-value mappings
+In plookup you can build key-value mappings as follows
+
+``` python
+def build_mapping():
+    keys = [1,3,5]
+    values = [2,4,6]
+
+    randomness = hash(keys, values)
+
+    mappings = []
+
+    for key , value in zip(keys,values):
+        mappings.append(key + randomness*value)
+    return(mappings)
+```
+
+
+It can be opened at the cost of 3 plookups
+
+``` python
+def open_mapping(mappings, keys, values):
+    randomness = hash(keys,values)
+    index = 1
+    # Prover can chose any mapping, key , value
+    mapping = plookup(mappings)
+    key = plookup(keys)
+    value = plookup(values)
```

Comment:
These are differnt plookup tables. They each have a single columns and the rows are the elements they contain. 

@han0110 am i getting the plookup/halo2 language correct ?

Reply:
If we follow the term described in [`halo2`](https://zcash.github.io/halo2/design/proving-system/lookup.html#generalizations), the thing we are looking up is a tuple `(key, value)` set. If we happen to organize the tuple as a 2-columns table with `key` as column1 and `value` as columns2, then it becomes like a table, each row in table represents a possible tuple.

When proving, halo2 helps to [`compress`](https://github.com/zcash/halo2/blob/27c4187673a9c6ade13fbdbd4f20955530c22d7f/src/plonk/lookup/prover.rs#L97) the 2-columns table with random linearly combination into a single-column table, which should be the `mapping` in the note, then it produces a subset argument.

So if the term **table** represents the tuple set we are looking up, there should only be one table here. The note is trying to emphasize that prover can't mix up different tuple `(key, value)` together like `(key_a, value_b)` because we assume a collision resistant `mapping`, is it correct @barryWhiteHat?

Reply:
Had a call with Barry. Yes, we want to prevent the keys to collide. Do you know a good way to express that in Halo2?